src/synchronizer.ts

```diff
@@ -0,0 +1,259 @@
+    const processedSnapshots = createProcessedSnapshotsComponent(components)
+
+    const stream = getDeployedEntitiesStreamFromSnapshots({ ...components, processedSnapshots }, options, snapshotsByServer)
+    logger.info('Starting to deploy entities from snapshots.')
+    for await (const entity of stream) {
+      // schedule the deployment in the deployer. the await DOES NOT mean that the entity was deployed entirely
+      // if the deployer is not synchronous. For example, the batchDeployer used in the catalyst just add it in a queue.
+      // Once the entity is truly deployed, it should call the method 'markAsDeployed'
+      await components.deployer.deployEntity({
+        ...entity,
+        markAsDeployed: async () => await processedSnapshots.entityProcessedFrom(entity.snapshotHash)
+      }, entity.servers)
+    }
+    logger.info('End deploying entities from snapshots.')
+
+    // Once the snapshots were correctly streamed, update the last entity timestamps
+    for (const [server, snapshots] of snapshotsByServer) {
+      increaseLastTimestamp(server, ...snapshots.map(s => s.lastIncludedDeploymentTimestamp))
+    }
+    // We only return servers that didn't fail to get its snapshots
+    return new Set(snapshotsByServer.keys())
+  }
+
+  async function bootstrapFromSnapshots() {
+    logger.debug(`Bootstrapping servers (snapshots): ${Array.from(bootstrappingServersFromSnapshots)}`)
+    const syncedServersFromSnapshot = await syncFromSnapshots(bootstrappingServersFromSnapshots)
+
+    for (const bootstrappedServer of syncedServersFromSnapshot) {
+      bootstrappingServersFromPointerChanges.add(bootstrappedServer)
+      bootstrappingServersFromSnapshots.delete(bootstrappedServer)
+    }
+  }
+
+  async function bootstrapFromPointerChanges() {
+    logger.debug(`Bootstrapping servers (Pointer Changes): ${Array.from(bootstrappingServersFromPointerChanges)}`)
+    const pointerChangesBootstrappingJobs = []
+    for (const bootstrappingServersFromPointerChange of bootstrappingServersFromPointerChanges) {
+      pointerChangesBootstrappingJobs.push(async () => {
+        try {
+          const lastEntityTimestamp = lastEntityTimestampFromSnapshotsByServer.get(bootstrappingServersFromPointerChange)
+          if (lastEntityTimestamp == undefined) {
+            throw new Error(`Can't start pointer changes stream without last entity timestamp for ${bootstrappingServersFromPointerChange}. This should never happen.`)
+          }
+          const fromTimestamp = lastEntityTimestamp - 20 * 60_000
+          await syncFromPointerChanges(bootstrappingServersFromPointerChange, { ...options, fromTimestamp, pointerChangesWaitTime: 0 }, () => false)
+          syncingServers.add(bootstrappingServersFromPointerChange)
+          bootstrappingServersFromPointerChanges.delete(bootstrappingServersFromPointerChange)
+        } catch (error) {
+          // If there's an error, the server doesn't pass to syncing state
```

Comment:
Should then throw an error?

Reply:
No, it runs bootstrap from pointer-changes for each catalyst. If one fails, it doesn't change its state from `bootstrapFromPointerChanges` to `syncingServers`. It'll be retried in the future